pkg/trace/sampler/probabilistic.go

```diff
@@ -0,0 +1,154 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2022-present Datadog, Inc.
+
+package sampler
+
+import (
+	"encoding/binary"
+	"encoding/hex"
+	"hash/fnv"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/proto/pbgo/trace"
+	"github.com/DataDog/datadog-agent/pkg/trace/config"
+	"github.com/DataDog/datadog-agent/pkg/trace/log"
+	"github.com/DataDog/datadog-agent/pkg/trace/watchdog"
+
+	"github.com/DataDog/datadog-go/v5/statsd"
+	"go.uber.org/atomic"
+)
+
+const (
+	// These constants exist to match the behavior of the OTEL probabilistic sampler.
+	// See: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/6229c6ad1c49e9cc4b41a8aab8cb5a94a7b82ea5/processor/probabilisticsamplerprocessor/tracesprocessor.go#L38-L42
+	numProbabilisticBuckets = 0x4000
+	bitMaskHashBuckets      = numProbabilisticBuckets - 1
+	percentageScaleFactor   = numProbabilisticBuckets / 100.0
+)
+
+// ProbabilisticSampler is a sampler that overrides all other samplers,
+// it deterministically samples incoming traces by a hash of their trace ID
+type ProbabilisticSampler struct {
+	enabled                  bool
+	hashSeed                 []byte
+	scaledSamplingPercentage uint32
+
+	statsd     statsd.ClientInterface
+	tracesSeen *atomic.Int64
+	tracesKept *atomic.Int64
+	tags       []string
+
+	// start/stop synchronization
+	stopOnce sync.Once
+	stop     chan struct{}
+	stopped  chan struct{}
+}
+
+// NewProbabilisticSampler returns a new ProbabilisticSampler that deterministically samples
+// a given percentage of incoming spans based on their trace ID
+func NewProbabilisticSampler(conf *config.AgentConfig, statsd statsd.ClientInterface) *ProbabilisticSampler {
+	hashSeedBytes := make([]byte, 4)
+	binary.LittleEndian.PutUint32(hashSeedBytes, conf.ProbabilisticSamplerHashSeed)
+	ps := &ProbabilisticSampler{
+		enabled:                  conf.ProbabilisticSamplerEnabled,
+		hashSeed:                 hashSeedBytes,
+		scaledSamplingPercentage: uint32(conf.ProbabilisticSamplerSamplingPercentage * percentageScaleFactor),
+		statsd:                   statsd,
+		tracesSeen:               atomic.NewInt64(0),
+		tracesKept:               atomic.NewInt64(0),
+		tags:                     []string{"sampler:probabilistic"},
+		stop:                     make(chan struct{}),
+		stopped:                  make(chan struct{}),
+	}
+	return ps
+}
+
+// Start starts up the ProbabilisticSamler's support routine, which periodically sends stats.
+func (ps *ProbabilisticSampler) Start() {
+	if !ps.enabled {
+		close(ps.stopped)
+		return
+	}
+	go func() {
+		defer watchdog.LogOnPanic(ps.statsd)
+		statsTicker := time.NewTicker(10 * time.Second)
```

Comment:
Wondering if the ticket timer is not configurable by design.

Reply:
I don't see a need to make this configurable.